File: services/analysis/chain-data/verification.js
/**
 * Etherscan contract verification checks (batched, rate-limit safe).
 */

import { ETHERSCAN_URLS, ETHERSCAN_KEY } from "../config.js";
import { fetchJson, buildEtherscanUrl, sleep } from "./http.js";

export async function enrichVerification(contractsRaw, chainId, batchSize = 5, delayMs = 250) {
  const base = ETHERSCAN_URLS[chainId] ?? ETHERSCAN_URLS[1];
  const enriched = [];

  for (let i = 0; i < contractsRaw.length; i += batchSize) {
    const batch = contractsRaw.slice(i, i + batchSize);

    for (const raw of batch) {
      let isVerified = false;
      try {
        const url = buildEtherscanUrl(base, {
          module: "contract",
          action: "getsourcecode",
          address: raw.contract_address,
          apikey: ETHERSCAN_KEY,
        });
        const data = await fetchJson(url);
        if (data.status === "1" && data.result?.length) {
          const src = data.result[0];
          isVerified = Boolean(
            src.SourceCode &&
            src.SourceCode !== "" &&
            src.SourceCode !== "1" &&
            src.ABI !== "Contract source code not verified",
          );
        }
      } catch (err) {
        console.debug(`[verification] Failed for ${raw.contract_address}: ${err.message}`);
      }

      enriched.push({ ...raw, is_verified: isVerified });
    }

    if (i + batchSize < contractsRaw.length) {
      await sleep(delayMs);
    }
  }

  return enriched;
}

File: services/indexer/src/verification.rs
use anyhow::Result;
use reqwest::Client;
use serde_json::Value;

use crate::config::{etherscan_base, Config};
use crate::deployments::{sleep_ms, RawDeployment};

pub async fn enrich_verification(
    cfg: &Config,
    client: &Client,
    chain_id: u32,
    rows: &[RawDeployment],
) -> Result<Vec<(RawDeployment, bool)>> {
    let base = etherscan_base(chain_id)?;
    let mut out = Vec::with_capacity(rows.len());

    for (i, row) in rows.iter().enumerate() {
        let verified = check_verified(client, cfg, base, &row.contract_address).await?;
        out.push((row.clone(), verified));

        if (i + 1) % 5 == 0 && i + 1 < rows.len() {
            sleep_ms(250).await;
        }
    }

    Ok(out)
}

pub async fn check_contract_verified(
    client: &Client,
    cfg: &Config,
    chain_id: u32,
    address: &str,
) -> Result<bool> {
    let base = etherscan_base(chain_id)?;
    check_verified(client, cfg, base, address).await
}

async fn check_verified(
    client: &Client,
    cfg: &Config,
    base: &str,
    address: &str,
) -> Result<bool> {
    let url = format!(
        "{base}?module=contract&action=getsourcecode&address={address}&apikey={}",
        cfg.etherscan_key
    );

    let resp: Value = client.get(&url).send().await?.json().await?;
    if resp.get("status").and_then(|v| v.as_str()) != Some("1") {
        return Ok(false);
    }

    let entry = resp
        .get("result")
        .and_then(|v| v.as_array())
        .and_then(|a| a.first());

    let Some(src) = entry else {
        return Ok(false);
    };

    let source = src.get("SourceCode").and_then(|v| v.as_str()).unwrap_or("");
    let abi = src.get("ABI").and_then(|v| v.as_str()).unwrap_or("");

    Ok(!source.is_empty() && source != "1" && abi != "Contract source code not verified")
}
